code/modules/ai/component_ai.dm

```diff
@@ -0,0 +1,86 @@
+/*
+Datums that represent an AI mind and it's way of doing various things like movement or handling a gun
+Base datums for stuff like humans or xenos have possible actions to do as well as attitudes
+*/
+
+//The most basic of AI; can pathfind to a turf and path around objects in it's path if needed to
+/datum/component/ai_behavior
+
+	var/turf/destinationturf //Turf that we want to get to
+	var/turf/lastturf //If this is the same as parentmob turf at HandleMovement() then we made no progress in moving, do HandleObstruction from there
+	var/obj/effect/AINode/current_node //Current node the parentmob is at
+	var/atom/atomtowalkto //Generic atom we're walking to; action states determine what type
+	var/obj/effect/AINode/destination_node
+	var/move_delay = 0 //The next world.time we can do a move at
+	var/datum/action_state/action_state //If we have an action state we feed it info and see what it tells us what to do
+	var/distance_to_maintain = 1 //Default distance to maintain from a target while in combat
+	var/datum/ai_mind/mind //Controls bsaic things like what to do once a action is completed or ability activations
+
+/datum/component/ai_behavior/Initialize(datum/ai_mind/mind_to_make)
+	if(!iscarbon(parent))
+		return COMPONENT_INCOMPATIBLE
+	if(!mind_to_make)
+		stack_trace("AI component was initialized without a mind to initialize parameter, stopping component creation.")
+		return COMPONENT_INCOMPATIBLE
+	var/atom/movable/parent2 = parent
+	SSai.aidatums += src
+	SSai_movement.lists_of_lists[1] += src
+	for(var/obj/effect/AINode/node in range(7))
+		if(node)
+			current_node = node
+			parent2.forceMove(current_node.loc)
+		break
+	if(!current_node)
+		stack_trace("An AI component was being attached to a movable atom however there's no nodes nearby; component removed.")
+		remove_everything()
+		return
+	mind = new mind_to_make(src)
+	mind.late_init()
+	RegisterSignal(parent, list(COMSIG_MOB_DEATH, COMSIG_PARENT_PREQDELETED ), .proc/remove_everything)
```

Comment:
parent_preqdeleted shouldnt need to be registered if you define a proper destroy

Reply:
then register mob_death to a wrapper proc that calls qdel(src)

Reply:
COSMIG_MOB_DEATH now calls qdel_self() that STOP_PROCESSING() then qdel(src)

COSMIG_PARENT_PREQDELETED has been removed